[WIP] flux-jobspec: first cut run command

trws opened the pull request:
This PR is a work in progress, but has the bones of the run interface discussed in #2213.  It produces some wrong jobspec right now, and doesn't process all cases, but it's enough that the structure is there and we can start playing with the way the arguments combine a bit.

SteVwonder commented:
Notes from the coffee time:

- Can you form a reasonable request where the slot is the memory (i.e., a memory-bound app wants to request 200GB per process and launch 100 processes)?
- How much memory do you get if you request just a single core?

SteVwonder commented:
Sorry @trws for leeting this linger so long 🙈 

Some feedback after poking around with it for a bit:

- It would be nice to support functionality to the `--export=NONE` in `jobspec srun`. Even if it's named differently or has different semantics.
- I think `shape` should be renamed to `slot-shape` to make connecting the shape argument to the slot concept easier.
- It might be easier to parse through the options if they were grouped using [add_argument_group](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_argument_group)
  - `--total-tasks`, `--tasks-per-slot`, and `--tasks-per-resource` could all be under `tasks`
  - `--nslots`, `--slot-shape`, `--shape-file`, could all be under `slot`
  - `--directory`, `--time`, `--export` could all be under `attributes`.
- The individual arguments could benefit from help descriptions.  Especially the `--shape` argument.  Having an example like `node/core[4]` would really help.  `--directory` is somewhat ambiguous given the `--file` argument, so a description would be helpful there too.

We may need to play with the `--shape` syntax some.  I got an error from zsh when I try and have multiple of a resource and I don't quote the argument:
```
→ flux jobspec run --shape node[2]/core --nslots 4 --tasks-per-slot 9 hostname | jq -S
zsh: no matches found: node[2]/core
```

SteVwonder commented:
>  It produces some wrong jobspec right now

Which inputs cause incorrect output?  The few simple cases that I tried seemed fine.

trws commented:
> Which inputs cause incorrect output? The few simple cases that I tried seemed fine.

None that I know of anymore, that is from an earlier version and I should have edited it out.

The current version builds various jobspec setups that flux can't currently read, but I don't know of any that are specification-wise invalid.

trws commented:
> Some feedback after poking around with it for a bit:
> 
> * It would be nice to support functionality to the `--export=NONE` in `jobspec srun`. Even if it's named differently or has different semantics.

Agreed.  I added something, let me know what you think.  This one works like this:

1. create initial dictionary either
  1. if nothing is specified about environment or --env-all is specified, initialize to current env
  2. initialize to an empty dict
2. add environment entries from the command-line, with repeatable --env=key[=val] or -e key[=val] arguments

Also considering having an `--env-file` argument that would be between step 1 and step 2, but did this for now.

> * I think `shape` should be renamed to `slot-shape` to make connecting the shape argument to the slot concept easier.

Agreed, done.

> * It might be easier to parse through the options if they were grouped using [add_argument_group](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_argument_group)
>   
>   * `--total-tasks`, `--tasks-per-slot`, and `--tasks-per-resource` could all be under `tasks`
>   * `--nslots`, `--slot-shape`, `--shape-file`, could all be under `slot`
>   * `--directory`, `--time`, `--export` could all be under `attributes`.

Did the first two, the others are currently job and environment.

> * The individual arguments could benefit from help descriptions.  Especially the `--shape` argument.  Having an example like `node/core[4]` would really help.  `--directory` is somewhat ambiguous given the `--file` argument, so a description would be helpful there too.

Agreed, added several.

> We may need to play with the `--shape` syntax some. I got an error from zsh when I try and have multiple of a resource and I don't quote the argument:
> 
> ```
> → flux jobspec run --shape node[2]/core --nslots 4 --tasks-per-slot 9 hostname | jq -S
> zsh: no matches found: node[2]/core
> ```

Yeah, we may need to poke at it some.  Zsh is probably trying to glob it and failing, not sure how best to deal with that without quoting, but will think on some alternatives.

grondo commented:
> Yeah, we may need to poke at it some. Zsh is probably trying to glob it and failing, not sure how best to deal with that without quoting, but will think on some alternatives.

You may be able to work around that with
```
setopt +o nomatch
```

The default in bash is nomatch globbing I think -- though even with nomatch we may have to train users to "quote early and often", because if there happens to be a file or files that *do* match the glob you can get unexpected results. (We run into this a lot with the ranged hostlist style expressions as well)

SteVwonder commented:
Thanks @trws!  This is looking really good.  That grouping looks slick.  And I like the composability of the environment arguments.

At some point, I think we will want to make `flux-run` itself a python script, much like we did for `flux-mini`, but I think that'll be easier once we have the Jobspec class in the bindings (rather than embedded in `flux-mini`).  So that can probably wait for a later PR.  In the meantime, do you mind tweaking the `flux-run` shell script so that `--help` is passed through and works?  Right now it punts it back to the user to do.
```
→ flux run -h
Usage: flux run OPTIONS command ...
(see flux-jobspec run for available OPTIONS)
```

---

The error message on `--file` could be improved since it technically can be combined with anything in the `job` and `environment` groups.

``` 
→ flux jobspec run --env-none --nslots 2 --dir ~/ --file t/jobspec/valid/basic.yaml hostname -h
flux-jobspec: ERROR: --file cannot be combined with other options

→ flux jobspec run --env-none --dir ~/ --file t/jobspec/valid/basic.yaml hostname -h | jq
{
  "tasks": [
    {
      "slot": "task",
      "count": {
        "per_slot": 1
      },
      "command": [
        "hostname",
        "-h"
      ],
      "attributes": {}
    }
  ],
  "attributes": {
    "system": {
      "environment": {},
      "duration": 3600,
      "cwd": "/home/herbein1/"
    }
  },
  "version": 1,
  "resources": [
    {
      "count": 1,
      "with": [
        {
          "count": 1,
          "type": "node"
        }
      ],
      "type": "slot",
      "label": "task"
    }
  ]
}
```

---

The other two big things that I'd like to see are:

1.  wire `flux-run` up to some tests (similar to t2007-mini-cmd.t)
2.  a man page for `flux-run` (although, maybe we push this off to a later PR when `flux-run` itself becomes a python script that itself supports the various arguments rather than shuttling them through `flux-jobspec`)

---

> You may be able to work around that with `setopt +o nomatch`

Thanks for the hint there @grondo.  That seemed to work for me!  I'm a little wary to enable that option globally for my environment though for the reason you listed.  I ran across a few horror stories while googling more about it.

SteVwonder commented on src/cmd/flux-jobspec.py:
```diff
@@ -176,6 +181,197 @@ def slurm_jobspec(args):
+        ):
+            logger.error("--file cannot be combined with other options")
+            sys.exit(1)
+        pass  # TODO< handle file
```
Lol.  I thought to try this functionality, but when it spit out a jobspec and didn't complain, I didn't even bother to double check the output was from the file I gave it :see_no_evil:

This should either get dropped or implemented before we merge (just making a comment so I don't click the button too quickly).

SteVwonder commented on src/cmd/flux-jobspec.py:
```diff
@@ -176,6 +181,180 @@ def slurm_jobspec(args):
+            yield res
+            res = res.get("with", None)
+        elif type(res) is list:
+            res = res[0]
```
Whilst working on the Jobspec class refactor, I noticed that this function only works on jobspec with lists of len == 1.  If any of the resources have multiple children, it won't traverse the extra children.

FWIW, the function I ended up with is:
```python
        def walk_helper(res_list):                                                                                                                                                                                                                                                                    
            for resource in res_list:                                                                                                                                                                                                                                                                 
                yield resource                                                                                                                                                                                                                                                                        
                children = resource.get("with", [])                                                                                                                                                                                                                                                   
                # TODO: convert to `yield from` after dropping 2.7                                                                                                                                                                                                                                    
                for x in walk_helper(children):                                                                                                                                                                                                                                                       
                    yield x
```

grondo commented:
@trws, ok to close this old PR? We could resubmit `flux run` as a new PR if needed.

trws commented:
Yeah, close it.  I think this should get resurrected but the code is now so out of date as to be a poor place to start from in many ways.  Thanks for the cleanup @grondo!
